### PR TITLE
Updated fw tags

### PR DESCRIPTION
**Pull request by kshelton**
Referencing #13035
- The longest tag that is a prefix of the instance name is used
- If any instance has a prefix tag, all instances must
- If no instances have a prefix tag, no tags are used

### CODE REVIEW

File: pkg/cloudprovider/providers/gce/gce.go [hunk 1]
```diff
@@ -450,16 +453,44 @@ func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, loadBalancerIP n
...
+		}
+		if len(longest_tag) > 0 {
+			tags.Insert(longest_tag)
+		} else if len(tags) > 0 {
+			return nil, fmt.Errorf("Some, but not all, instances have prefix tags (%s is missing)", instance.Name)
+		}
+	}
+
+	if len(tags) == 0 {
+		glog.V(2).Info("No instances had tags, creating rule without target tags")
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/gce/gce.go]: We should return an error in this case.

File: pkg/cloudprovider/providers/gce/gce.go [hunk 2]
```diff
@@ -450,16 +453,44 @@ func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, loadBalancerIP n
...
-func (gce *GCECloud) computeHostTag(host string) string {
-	host = strings.SplitN(host, ".", 2)[0]
-	lastHyphen := strings.LastIndex(host, "-")
-	if lastHyphen == -1 {
-		return host
+// We grab all tags from all instances being added to the pool.
+// * The longest tag that is a prefix of the instance name is used
+// * If any instance has a prefix tag, all instances must
+// * If no instances have a prefix tag, no tags are used
+func (gce *GCECloud) computeHostTags(hosts []string) ([]string, error) {
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/gce/gce.go]: Why would there ever be more than one tag?

File: pkg/cloudprovider/providers/gce/gce.go (hunk 1 above)
Comment (kshelton) [inline pkg/cloudprovider/providers/gce/gce.go]: See  #14430, creating the rule without a tags should not be an error case if none of the instances have tags.

File: pkg/cloudprovider/providers/gce/gce.go (hunk 2 above)
Comment (kshelton) [inline pkg/cloudprovider/providers/gce/gce.go]: See  #14430, only in the current GKE case can you assume that all instances will be tagged the same.  This allows for setting separate tags per instance and when the pool logic only adds the instances serving the pods to the pool, you will end up with only those instances allowed to receive that traffic.

**Comment by brendandburns** [comment]
@k8s-bot test this please

Ok, I think I'm convinced that this is ok.

Please add unit tests.

File: pkg/cloudprovider/providers/gce/gce.go (hunk 2 above)
Comment (brendandburns) [inline pkg/cloudprovider/providers/gce/gce.go]: Ok, I'm convinced.   Please add unit tests for this function.

**Comment by kshelton** [comment]
I don't relive unit tests can be easily made for this, as it calls in to the GCE API (to retrieve the tags) and there is no mocking for this in the current unit tests.  I imagine this is why a lot of features are not present in the GCE provider unit tests.  I can do e2e tests if desired.  

Alternatively, I can move the getting of the tags to a separate call, or the parent, and pass in the list to the function and provide a unit test for the logic in computeHostTags.

**Comment by brendandburns** [comment]
LGTM

**Pull request merged**